### PR TITLE
Bosch poc

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs [hunk 1]
```diff
@@ -772,6 +786,17 @@ public void Prune(CancellationToken cancellationToken)
...
+                        if (connectionInfo.IsInUse)
+                        {
+                            if (_connectionPoolDiagnosticEventHandler != null)
+                            {
+                                _connectionPoolDiagnosticEventHandler.Invoke(
+                                    new DiagnosticEvent($"Removing connection in use is triggered. ConnectionId: {connection.ConnectionId}"));
+                            }
+                            _connectionsInUse.Remove(connection);
+                        }
+
```
Comment (JamesKovacs) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs:792]: We have a race condition here. We could take the connections lock on line 761 and find that certain connections were in use and others weren't. We then repeatedly reacquire the lock on line 778. While the lock isn't held, connections could move from the pool to in use or vice versa. Thus the logic between lines 787 and 798 could be incorrect.

I would suggest that lines 787 to 798 should look something like this instead:

```
RemoveConnection(connection);
_connections.Remove(connection);
var inUse = _connectionsInUse.Remove(connection);
if (inUse && _connectionPoolDiagnosticEventHandler != null)
{
    _connectionPoolDiagnosticEventHandler.Invoke(
        new DiagnosticEvent($"Removing connection in use is triggered. ConnectionId: {connection.ConnectionId}"));
}
```

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs [hunk 2]
```diff
@@ -744,22 +749,31 @@ public void Clear()
...
+                    _connectionsInUse.Clear();
 
                     SignalOrReset();
                 }
             }
 
             public void Prune(CancellationToken cancellationToken)
             {
-                PooledConnection[] expiredConnections;
+                (PooledConnection Connection, bool IsInUse)[] expiredConnections;
```
Comment (JamesKovacs) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs]: We don't need the `IsInUse` property. See below.

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs [hunk 3]
```diff
@@ -808,6 +834,7 @@ public void Return(PooledConnection connection)
                 lock (_lock)
                 {
                     _connections.Add(connection);
+                    _connectionsInUse.RemoveAll(c => c.ConnectionId == connection.ConnectionId);
```
Comment (JamesKovacs) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs]: Why isn't this `_connectionsInUse.Remove(connection);`?

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Found a race condition, which I think I have an easy fix for. See my suggestion.

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs]: true, it should be

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs]: Done

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs:792]: true, done

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs]: done

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request closed without merging**